src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -823,6 +825,46 @@
     (sutils/sql-hash-as-str "hash") certname]
    (comp :catalog first sql/result-set-seq)))
 
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Catalog inputs updates/changes
+
+(defn add-certname-id-to-catalog-inputs [certid inputs]
+  (map (fn [input] (into [certid] input)) inputs))
+
+(defn delete-catalog-inputs! [certid]
+  (jdbc/delete! :catalog_inputs ["certname_id = ?" certid]))
+
+(defn store-catalog-inputs!
+  [certid inputs]
+    (jdbc/insert-multi!
+      :catalog_inputs
+      [:certname_id :type :name]
+      (add-certname-id-to-catalog-inputs certid inputs)))
+
+(defn update-catalog-input-metadata!
+  [certid catalog-uuid last-updated]
+  (jdbc/update! :certnames
+                {:catalog_inputs_uuid catalog-uuid :catalog_inputs_timestamp last-updated}
+                ["id = ?" certid]))
+
+(pls/defn-validated replace-catalog-inputs!
+  [certname :- s/Str
+   catalog_uuid :- s/Str
+   inputs :- [[s/Str]]
+   updated :- pls/Timestamp]
+  (jdbc/with-db-transaction []
+    (let [updated (to-timestamp updated)
+          catalog-uuid (sutils/munge-uuid-for-storage catalog_uuid)
+          certid (-> (jdbc/query
+                       (hcore/format {:select [:id]
+                                      :from [:certnames]
+                                      :where [:= :certname certname]}))
+                     first
+                     :id)]
+      (delete-catalog-inputs! certid)
+      (store-catalog-inputs! certid inputs)
```

Comment:
Perhaps a candidate for an upsert (on conflict replace) once we're ready to pursue them more enthusiastically.  Wonder if we might want to leave a comment (something greppable) to make it easier to find when we're working on that, if we're not ready to do it now.